### PR TITLE
Indexing via discrimination tree

### PR DESCRIPTION
**Pull request by FissoreD**

### CODE REVIEW

**Comment by gares** [comment]
Looks much better than this morning. Do the tests pass if we always use the DT index?

**Comment by FissoreD** [comment]
Most tests pass, but not all yet. I should look deeper to spot the bug. 
There are some  slow down in performance especially for tests like `crypt`.
Even if I think this particular test is not a good use case for indexing with trie, at least in the current implementation 
where we do not limit the height/depth of the path representing the goal, I suppose
lot of time is spent during the creation of the path corresponding to the goal.
There are 5 timeout:
```
TIMEOUT  lyp (type checker for λΥP)                elpi
TIMEOUT  map (defined list) (stdlib map reference)   elpi
TIMEOUT  map (rbtree) (stdlib map)                   elpi
TIMEOUT  mu (standard Prolog benchmark)              elpi
TIMEOUT  set (stdlib set)                            elpi
```
And three failures:
```
KO       eta_as (eta expansion of as clause)         elpi
KO       trace-browser-chr (trace generation)        elpi
KO       uvar_keyword (uvar kwd status at the meta level) elpi
```

**Comment by gares** [comment]
Indeed, bounding the index to the given depth seems necessary

**Comment by gares** [comment]
The failures are corner cases, we can look into them together. Bounding the depth of the trie seems more important

File: Makefile [hunk 1]
```diff
@@ -134,4 +134,4 @@ menhir-complete-errormsgs:
 menhir-strip-errormsgs:
 	sed -e "/^##/d" -i.bak src/parser/error_messages.txt
 
-.PHONY: tests help install build clean gh-pages
+.PHONY: tests help install build clean gh-pages
```
Comment (gares) [inline Makefile:137]:
```suggestion
.PHONY: tests help install build clean gh-pages

```

File: CHANGES.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 Library:
   - New `std.fold-right`
+  - New clause retrieval through discrimination tree where the index of all the
+    parameters are set to `0` but one: the argument of which the search is done
```
Comment (gares) [inline CHANGES.md:6]:
```suggestion
  - New clause retrieval through discrimination tree. This new index is enabled
    whenever the `:index` directive selects only one argument  with a depth `> 1`.
```

File: src/compiler.ml [hunk 3]
```diff
@@ -2406,7 +2406,8 @@ let run
       let mode = try C.Map.find name modes with Not_found -> [] in
       let declare_index, index =
         match tindex with
-        | Some (Ast.Structured.Index l) -> true, chose_indexing state name l
+        | Some (Ast.Structured.Index l) ->
+          true, chose_indexing state name l
```
Comment (gares) [inline src/compiler.ml:2410]:
```suggestion
        | Some (Ast.Structured.Index l) -> true, chose_indexing state name l
```

File: src/data.ml [hunk 4]
```diff
@@ -115,6 +115,80 @@ let uvar_isnt_a_blocker { uid_private } = uid_private > 0 [@@inline];;
...
+    | Constant (_,a) -> a 
+    | Variable | PrimitiveType _ -> 0
+
+  let skip (path: path) : path =
+    let rec aux arity path = 
+    if arity = 0 then path else match path with 
+      | [] -> assert false 
+      | m::tl -> aux (arity-1+arity_of m) tl in 
+    match path with
+      | [] -> failwith "Skipping empty path is not possible"
```
Comment (gares) [inline src/data.ml:176]:
```suggestion
      | [] -> anomaly "Skipping empty path is not possible"
```

File: src/data.ml [hunk 5]
```diff
@@ -167,9 +238,20 @@ type suspended_goal = {
   goal : int * term
 }
 
+(** 
+  Used to index the parameters of a predicate P
+  - [MapOn N] -> Indexing is done by unifying the Nth parameter of P with the
+                 query
```
Comment (gares) [inline src/data.ml:244]:
```suggestion
  - [MapOn N] -> N-th argument at depth 1 (head symbol only)
```

File: src/data.ml [hunk 6]
```diff
@@ -167,9 +238,20 @@ type suspended_goal = {
...
+  - [Hash L]  -> L is the list of depths given by the urer for the parameters of
+                 P. Indexing is done by hashing all the parameters with a non
+                 zero depth and comparing it with the hashing of the parameters
+                 of the query
+  - [IndexWithTrie N] -> Indexing is done on the Nth parameter using tries
```
Comment (gares) [inline src/data.ml:249]:
```suggestion
  - [IndexWithTrie N] -> N-th argument at arbitrary depth (TODO bound)
```

File: src/discrimination_tree.ml [hunk 7]
```diff
@@ -0,0 +1,146 @@
+(* elpi: embedded lambda prolog interpreter                                  *)
+(* license: GNU Lesser General Public License Version 2.1 or later           *)
+(* ------------------------------------------------------------------------- *)
+
+module type IndexableTerm = sig
+  type cell
+  type path = cell list
+
+  val show_cell : cell -> string
+  val pp_cell : Format.formatter -> cell -> unit
```
Comment (gares) [inline src/discrimination_tree.ml:10]:
```suggestion
  include Show with type t := cell
```

File: src/discrimination_tree.ml [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+  let pp = Trie.pp D.pp
+  let show = Trie.show D.pp
+  let empty = Trie.empty
+  let iter dt f = Trie.iter (fun p x -> f p x) dt
+  let fold dt f = Trie.fold (fun p x -> f p x) dt
+
+  let index tree ps info =
+    let ps_set = try Trie.find ps tree with Not_found -> [] in
+    Trie.add ps (info :: ps_set) tree
```
Comment (gares) [inline src/discrimination_tree.ml:91]: Maybe Trie.add could have this semantics  already (modulo having the trie be on 'a list all the times).

File: src/discrimination_tree.ml [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+  let remove_index tree ps info =
+    try
+      let ps_set = List.filter (( = ) info) (Trie.find ps tree) in
```
Comment (gares) [inline src/discrimination_tree.ml:95]:
```suggestion
      let ps_set = List.filter (D.equal info) (Trie.find ps tree) in
```
sine `=` may be too precise for `D.elt` (e.g. if there is a quotient)

File: src/discrimination_tree.ml [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+      if ps_set = [] then Trie.remove ps tree else Trie.add ps ps_set tree
```
Comment (gares) [inline src/discrimination_tree.ml:96]: I think this can go to the trie API as well.

File: src/runtime.ml [hunk 11]
```diff
@@ -2560,9 +2617,24 @@ let get_clauses ~depth predicate goal { index = m } =
...
        let cl = List.flatten (Ptmap.find_unifiables hash args_idx) in
        List.(map fst (sort (fun (_,cl1) (_,cl2) -> cl2 - cl1) cl))
+     | IndexWithTrie {argno; mode; args_idx} -> 
+        let mode_arg = nth_not_bool_default mode argno in 
+        let arg = arg_to_trie_path ~depth (trie_goal_args goal argno) in
+        [%spy "dev:disc-tree-filter-number1" ~rid 
+          pp_string "Current path is" (pp_path pp_int) arg
+          pp_string " and current DT is " DT.pp args_idx];
+        (* TODO: check better this bool of the condition... *)
+        let unifying_clauses = if false && mode_arg then
```
Comment (gares) [inline src/runtime.ml:2627]: ??

File: src/trie.mli [hunk 12]
```diff
@@ -0,0 +1,19 @@
+module Make : functor (M : Elpi_util.Util.Map.S) -> sig
```
Comment (gares) [inline src/trie.mli:1]: put the same license as in the .ml

**Comment by gares** [comment]
many changes are nitpicks to reduce the diff

File: src/data.ml [hunk 13]
```diff
@@ -181,6 +181,7 @@ and type t = (clause * int) list = struct
   type t = elt list [@@deriving show]
   
   let get_time_stamp = snd
+  let equal a b = a = b
```
Comment (gares) [inline src/data.ml:184]: I think you should use the timestamp as a unique id, here you are comparing the clause "code" which can be large.

Comment (FissoreD) [reply src/data.ml:184]: I totally agree

File: src/runtime.ml [hunk 14]
```diff
@@ -2463,7 +2463,7 @@ and arg_to_trie_path ~safe ~depth t path_depth : Discrimination_tree.path =
     | Cons (x,xs) ->
       let hd_path = arg_to_trie_path ~safe ~depth x path_depth in
       let tl_path = arg_to_trie_path ~safe ~depth xs path_depth in
-      mkConstant ~safe Global_symbols.consc 2 :: hd_path @ tl_path
+      mkConstant ~safe Global_symbols.consc (if path_depth = 0 then 0 else 2) :: hd_path @ tl_path
```
Comment (gares) [inline src/runtime.ml:2466]: Then I guess it is not just cons to be affected

Comment (FissoreD) [reply src/runtime.ml:2466]: For the other cases (app and builtins) where the arity could be > 0, we alredy do this check

File: src/data.ml [hunk 15]
```diff
@@ -115,6 +116,32 @@ let uvar_isnt_a_blocker { uid_private } = uid_private > 0 [@@inline];;
...
+
+let to_mode = function true -> Input | false -> Output
+
+(* Simpler pretty printer for clause *)
+let pp_clause_simple (fmt:Format.formatter) (cl: clause) = Format.fprintf fmt "clause" (*
+  Format.fprintf fmt "[clause_args:";
+  pplist pp_term ", " fmt cl.args;
+  Format.fprintf fmt " ;; clause_hyps:";
+  pplist pp_term ", " fmt cl.hyps;
+  Format.fprintf fmt "]";*)
```
Comment (gares) [inline src/data.ml:141]:
```suggestion
```

File: src/data.ml [hunk 16]
```diff
@@ -167,9 +192,19 @@ type suspended_goal = {
...
 }
 
+(** 
+  Used to index the parameters of a predicate P
+  - [MapOn N] -> N-th argument at depth 1 (head symbol only)
+  - [Hash L]  -> L is the list of depths given by the urer for the parameters of
+                 P. Indexing is done by hashing all the parameters with a non
+                 zero depth and comparing it with the hashing of the parameters
+                 of the query
+  - [IndexWithTrie N] -> N-th argument at arbitrary depth (TODO bound)
```
Comment (gares) [inline src/data.ml:202]:
```suggestion
  - [IndexWithTrie N D] -> N-th argument at D depth
```

File: src/discrimination_tree.ml [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+(* license: GNU Lesser General Public License Version 2.1 or later           *)
+(* ------------------------------------------------------------------------- *)
+let arity_bits = 4
+let k_bits = 2
+
+(* value , arity, k *)
+let kConstant = 0 (* (constant << arity_bits) lor arity *)
+let kPrimitive = 1 (*Elpi_util.Util.CData.t hash *)
+let kVariable = 2
+let kOther = 3
```
Comment (gares) [inline src/discrimination_tree.ml:11]:
```suggestion
(* 4 constructors encoded as: arg_value , arity, kno *)
let kConstant = 0
let kPrimitive = 1
let kVariable = 2
let kOther = 3

```

File: src/discrimination_tree.ml [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+  else if k == kPrimitive then Format.fprintf fmt "Primitive"
+  else Format.fprintf fmt "%o" k
+
+let show_cell n =
+  Format.asprintf "%a" pp_cell n
+
+module Trie = struct
+  (*
+   * Trie: maps over lists.
+   * Copyright (C) 2000 Jean-Christophe FILLIATRE
```
Comment (gares) [inline src/discrimination_tree.ml:58]:
```suggestion
   * Note: This code is a heavily modified version of the original code.
   * Copyright (C) 2000 Jean-Christophe FILLIATRE
```

**Review by gares — APPROVED** [review]

**Pull request merged**